### ISSUE TITLE
Add ! prefix to "set" topics

### ISSUE THREAD

**Issue by oeysteinhansen** [homieiot/convention#44]
Attributes topics have the "$" prefix.
I would like us to prefix the command topics, e.g. "set", with "!" or another appropriate character. 

The reason is to make it easier to distinguish the calls going to the device from the controller.
Distinguish between state updates and commands.
(Command topics will not have the Retain bit set.)

Examples:

1. Setting light intensity from the controller to 50%
 - Controller publish
    CTR: homie/686f6d6965/light/intensity/!set "50"
 - Device receives, and applies the change, then sends a property update
    DEV: homie/686f6d6965/light/intensity "50"

2. Update device publishing interval to every 2.0 seconds
 - Controller publish
    CTR: homie/686f6d6965/$interval/!set "2.0"
 - Device receives, and applies the change, then sends a attribute update
    DEV: homie/686f6d6965/$interval "2.0"

3. Reboot device (implementation specific)
 - Controller publish
    CTR: homie/686f6d6965/$implementation/!reboot ""
 - Device reboots

**Comment by ThomDietrich** [comment]
Hello @oeysteinhansen,
MQTT topic names normally consist of words/letters and numbers. It is highly unusual to use special characters like a "!". The dollar sign has a special meaning, see: http://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices

The Homie convention already relies on the idea of speaking keywords. I feel the word "set" or "reboot" already imply a command.
Secondly the Homie convention defines how devices can communicate with each other. A certain knowledge is needed to be able to interact using the convention. This includes the topics structure and commands. The fact that the topic "set" is to be used by the counterpart for command messages is already known at this point.
With this scenario in mind, adding an additional character doesn't make a difference.

So in conclusion, I don't think that an unusual character added to command topic names will improve the Homie convention.

May I ask where your idea originates from? Maybe a related aspect of the Homie convention is misleading?

**Comment by oeysteinhansen** [comment]
I had several thoughts in mind when suggesting the prefix. I did not want to jam everything into this issue, but I can describe some of  them her. 

I can see that the mechanism for detecting nodes and properties are by a separate attributes $nodes and $properites. I would introduce that all topic not starting with a special character ($ or ! ) is the ID of the node / property. Eliminating the need for the $nodes and $properties attributes.

I would introduce that a node / property does not directly have a value, instead all values are represented as attributes. 
Example: 
homie/super-car/engine/$name → "Car engine" 
homie/super-car/engine/$type → "V8" 
homie/super-car/engine/speed/$value → "3.7"
homie/super-car/engine/direction/$value → "0.15"
homie/super-car/engine/temperature/$value → "21.5"

Future expansions:
Nodes and Properties would be equal. Both having describing attributes.
Nodes could have sub-nodes, allowing adding gateways, or domains.
Array of Nodes can be represented as sub-nodes instead of the _INDEX postfix of the NodeID. 
Example: 
homie/super-car/lights/$name → "Lights"
homie/super-car/lights/$type → "array"
homie/super-car/lights/0/$name → "Back lights"
homie/super-car/lights/0/intensity/$name → "Intensity"
homie/super-car/lights/0/intensity/$value → "0"
homie/super-car/lights/1/$name → "Front lights"
homie/super-car/lights/1/intensity/$name → "Intensity"
homie/super-car/lights/1/intensity/$value → "100"

**Comment by marvinroger** [comment]
@oeysteinhansen I understand the concern, but unfortunately we have to freeze the evolution at some point. The 2.1.0 is not released yet, and if we never stop changing things, it will never be. So let's put that aside for now. 😉

**Comment by ThomDietrich** [comment]
That is why an active maintainer group is important. I did criticize the idea and think it is important to get a second opinion (yes, I can be wrong). With @marvinroger 's and your response this is in fact an issue we can reject.
@oeysteinhansen thanks for your input!

**Comment by davidgraeff** [comment]
@oeysteinhansen: Thank you for your input :)

I hope you don't feel offended. I'm worried about the maintainership of this project and picked your issue as reference for one of many stale issues. You can follow and comment the discussion in the open pull request :)
